### PR TITLE
Make redis backend greenlet-safe

### PR DESCRIPTION
**Pull request by akhalyavka**
Partially addresses https://github.com/celery/celery/issues/4363 https://github.com/celery/celery/issues/4670

See my comment in https://github.com/celery/celery/pull/5145#issuecomment-488873389 for explanation.
This PR doesn't completely solve the problem, but at least it should solve it for gevent/eventlet.

### CODE REVIEW

File: celery/backends/redis.py [hunk 1]
```diff
@@ -71,20 +73,22 @@ def drain_events(self, timeout=None):
 
     def consume_from(self, task_id):
         if self._pubsub is None:
-            return self.start(task_id)
+            return self.start()
```
Comment (yevhen-m) [inline celery/backends/redis.py]: Why remove `task_id`?

Comment (akhalyavka) [reply celery/backends/redis.py]: got confused at some point. you are right, there's no reason

File: celery/backends/redis.py [hunk 2]
```diff
@@ -44,13 +44,13 @@
 class ResultConsumer(async.BaseResultConsumer):
 
     _pubsub = None
+    _subscribe_lock = threading.Lock()
```
Comment (yevhen-m) [inline celery/backends/redis.py]: Why have you moved the lock here?

Comment (yevhen-m) [reply celery/backends/redis.py]: Looks like this lock should manage access to self._pubsub, that's why it should be on the class level.

File: celery/backends/redis.py [hunk 3]
```diff
@@ -76,15 +79,17 @@ def consume_from(self, task_id):
 
     def _consume_from(self, task_id):
         key = self._get_key_for_task(task_id)
-        if key not in self.subscribed_to:
-            self.subscribed_to.add(key)
-            self._pubsub.subscribe(key)
+        with self._subscribe_lock:
+            if key not in self.subscribed_to:
+                self.subscribed_to.add(key)
+                self._pubsub.subscribe(key)
```
Comment (yevhen-m) [inline celery/backends/redis.py]: It seems to me that only `self._pubsub.subscribe(key)` should be be behind a lock.

File: celery/backends/redis.py (hunk 2 above)
Comment (akhalyavka) [reply celery/backends/redis.py]: Yes, correct

File: celery/backends/redis.py (hunk 3 above)
Comment (akhalyavka) [reply celery/backends/redis.py]: It's usually a good practice to guard related conditions with a lock, however here it doesn't really matter.
as far as i don't have a strong opinion about placing of this lock, i'll put it as close as possible

**Pull request merged**